Constant used in attribute on local function warns that it is unused

**Version Used**: Compiler version: '3.8.0-4.20460.4 (64f2392e)'. Language version: 9.0.

**Steps to Reproduce**:

Compile the following code: ([SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEAMACFBGAdAJQFcA7GASzDlwElTEB7ABwGVEA3MgYzigG4BYAFBCUAZiwAmDABUeMIQG8hGFVnEoALDLkBZODAAW9ACYAKAJTLVSwarsZO9YrCzZMxgDYeAchEoYAXgwAIllYXE8PYIFbexUrOIBtABEvajBGegQYU0jfSnMAXQT7HAA2DDgAD3gEYiwtX3I2OBQAVnInCxi4+Ni4prIW9s7ibpKAXyEJoA==))

```csharp
using System.Runtime.InteropServices;

public class Test
{
    public void TestMethod()
    {
        const string dllName = "Test.dll";
        
        [DllImport(dllName)]
        static extern void NativeFunction();
        
        NativeFunction();
    }
}
```

**Expected Behavior**: No warnings or errors.

**Actual Behavior**:

The compiler warns that `dllName` is not used:

```
warning CS0219: The variable 'dllName' is assigned but its value is never used
```

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
@@ -1686,11 +1686,10 @@ namespace Microsoft.CodeAnalysis.CSharp
                 NoteWrite(parameter, value: null, read: true);
             }
 
-            if (parameter is SourceComplexParameterSymbolBase { ContainingSymbol: LocalFunctionSymbol or LambdaSymbol } sourceComplexParam &&
-                sourceComplexParam.BindParameterAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
+            if (parameter is SourceComplexParameterSymbolBase { ContainingSymbol: LocalFunctionSymbol or LambdaSymbol } sourceComplexParam)
             {
                 // Mark attribute arguments as used.
-                VisitAttributes(boundAttributes);
+                VisitAttributes(sourceComplexParam.BindParameterAttributes());
             }
         }
 
@@ -1699,6 +1698,11 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// </summary>
         private void VisitAttributes(ImmutableArray<(CSharpAttributeData, BoundAttribute)> boundAttributes)
         {
+            if (boundAttributes.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
             foreach (var (attributeData, boundAttribute) in boundAttributes)
             {
                 // Skip invalid attributes (e.g., with a non-constant argument) to avoid superfluous diagnostics.
@@ -1917,10 +1921,7 @@ namespace Microsoft.CodeAnalysis.CSharp
                     if (stmt is BoundLocalFunctionStatement localFunctionStatement)
                     {
                         // Mark attribute arguments as used.
-                        if (localFunctionStatement.Symbol.BindMethodAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
-                        {
-                            VisitAttributes(boundAttributes);
-                        }
+                        VisitAttributes(localFunctionStatement.Symbol.BindMethodAttributes());
 
                         VisitAlways(stmt);
                     }
@@ -2147,10 +2148,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             this.CurrentSymbol = node.Symbol;
 
             // Mark attribute arguments as used.
-            if (node.Symbol.BindMethodAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
-            {
-                VisitAttributes(boundAttributes);
-            }
+            VisitAttributes(node.Symbol.BindMethodAttributes());
 
             var oldPending = SavePending(); // we do not support branches into a lambda
 

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
@@ -1697,10 +1697,16 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <summary>
         /// Marks attribute arguments as used.
         /// </summary>
-        private void VisitAttributes(ImmutableArray<BoundAttribute> boundAttributes)
+        private void VisitAttributes(ImmutableArray<(CSharpAttributeData, BoundAttribute)> boundAttributes)
         {
-            foreach (var boundAttribute in boundAttributes)
+            foreach (var (attributeData, boundAttribute) in boundAttributes)
             {
+                // Skip invalid attributes (e.g., with a non-constant argument) to avoid superfluous diagnostics.
+                if (attributeData.HasErrors)
+                {
+                    continue;
+                }
+
                 foreach (var attributeArgument in boundAttribute.ConstructorArguments)
                 {
                     VisitRvalue(attributeArgument);

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs
@@ -1690,16 +1690,24 @@ namespace Microsoft.CodeAnalysis.CSharp
                 sourceComplexParam.BindParameterAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
             {
                 // Mark attribute arguments as used.
-                foreach (var boundAttribute in boundAttributes)
+                VisitAttributes(boundAttributes);
+            }
+        }
+
+        /// <summary>
+        /// Marks attribute arguments as used.
+        /// </summary>
+        private void VisitAttributes(ImmutableArray<BoundAttribute> boundAttributes)
+        {
+            foreach (var boundAttribute in boundAttributes)
+            {
+                foreach (var attributeArgument in boundAttribute.ConstructorArguments)
                 {
-                    foreach (var attributeArgument in boundAttribute.ConstructorArguments)
-                    {
-                        VisitRvalue(attributeArgument);
-                    }
-                    foreach (var attributeNamedArgumentAssignment in boundAttribute.NamedArguments)
-                    {
-                        VisitRvalue(attributeNamedArgumentAssignment.Right);
-                    }
+                    VisitRvalue(attributeArgument);
+                }
+                foreach (var attributeNamedArgumentAssignment in boundAttribute.NamedArguments)
+                {
+                    VisitRvalue(attributeNamedArgumentAssignment.Right);
                 }
             }
         }
@@ -1900,8 +1908,14 @@ namespace Microsoft.CodeAnalysis.CSharp
                 // First phase
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind == BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunctionStatement)
                     {
+                        // Mark attribute arguments as used.
+                        if (localFunctionStatement.Symbol.BindMethodAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
+                        {
+                            VisitAttributes(boundAttributes);
+                        }
+
                         VisitAlways(stmt);
                     }
                 }
@@ -2125,6 +2139,12 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
             var oldSymbol = this.CurrentSymbol;
             this.CurrentSymbol = node.Symbol;
+
+            // Mark attribute arguments as used.
+            if (node.Symbol.BindMethodAttributes() is { IsDefaultOrEmpty: false } boundAttributes)
+            {
+                VisitAttributes(boundAttributes);
+            }
 
             var oldPending = SavePending(); // we do not support branches into a lambda
 

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
@@ -575,7 +575,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
         /// <summary>
         /// Binds attributes applied to this parameter.
         /// </summary>
-        public ImmutableArray<BoundAttribute> BindParameterAttributes()
+        public ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindParameterAttributes()
         {
             return BindAttributes(GetAttributeDeclarations(), WithTypeParametersBinderOpt);
         }

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs
@@ -577,23 +577,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
         /// </summary>
         public ImmutableArray<BoundAttribute> BindParameterAttributes()
         {
-            var binder = new ContextualAttributeBinder(WithTypeParametersBinderOpt, this);
-            var boundAttributeArrayBuilder = ArrayBuilder<BoundAttribute>.GetInstance();
-            foreach (var attributeListSyntaxList in GetAttributeDeclarations())
-            {
-                foreach (var attributeListSyntax in attributeListSyntaxList)
-                {
-                    foreach (var attributeSyntax in attributeListSyntax.Attributes)
-                    {
-                        var boundType = binder.BindType(attributeSyntax.Name, BindingDiagnosticBag.Discarded);
-                        var boundTypeSymbol = (NamedTypeSymbol)boundType.Type;
-                        var boundAttribute = new ExecutableCodeBinder(attributeSyntax, binder.ContainingMemberOrLambda, binder)
-                            .BindAttribute(attributeSyntax, boundTypeSymbol, this, BindingDiagnosticBag.Discarded);
-                        boundAttributeArrayBuilder.Add(boundAttribute);
-                    }
-                }
-            }
-            return boundAttributeArrayBuilder.ToImmutableAndFree();
+            return BindAttributes(GetAttributeDeclarations(), WithTypeParametersBinderOpt);
         }
 
         internal override void EarlyDecodeWellKnownAttributeType(NamedTypeSymbol attributeType, AttributeSyntax attributeSyntax)

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
@@ -368,7 +368,6 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
         /// </summary>
         public ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindMethodAttributes()
         {
-            Debug.Assert(OuterBinder is not null);
             return BindAttributes(GetAttributeDeclarations(), OuterBinder);
         }
 #nullable disable

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
@@ -362,6 +362,15 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
 
             return base.EarlyDecodeWellKnownAttribute(ref arguments);
         }
+
+        /// <summary>
+        /// Binds attributes applied to this method.
+        /// </summary>
+        public ImmutableArray<BoundAttribute> BindMethodAttributes()
+        {
+            Debug.Assert(OuterBinder is not null);
+            return BindAttributes(GetAttributeDeclarations(), OuterBinder);
+        }
 #nullable disable
 
         public override bool AreLocalsZeroed

--- a/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
@@ -366,7 +366,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
         /// <summary>
         /// Binds attributes applied to this method.
         /// </summary>
-        public ImmutableArray<BoundAttribute> BindMethodAttributes()
+        public ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindMethodAttributes()
         {
             Debug.Assert(OuterBinder is not null);
             return BindAttributes(GetAttributeDeclarations(), OuterBinder);

--- a/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
@@ -517,6 +517,30 @@ namespace Microsoft.CodeAnalysis.CSharp
                 return d.HasLazyInfo ? d.LazyInfo is LazyObsoleteDiagnosticInfo : d.Info.IsObsoleteDiagnostic();
             }
         }
+
+        /// <summary>
+        /// Binds attributes applied to this symbol.
+        /// </summary>
+        protected ImmutableArray<BoundAttribute> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder rootBinder)
+        {
+            var binder = new ContextualAttributeBinder(rootBinder, this);
+            var boundAttributeArrayBuilder = ArrayBuilder<BoundAttribute>.GetInstance();
+            foreach (var attributeListSyntaxList in attributeDeclarations)
+            {
+                foreach (var attributeListSyntax in attributeListSyntaxList)
+                {
+                    foreach (var attributeSyntax in attributeListSyntax.Attributes)
+                    {
+                        var boundType = binder.BindType(attributeSyntax.Name, BindingDiagnosticBag.Discarded);
+                        var boundTypeSymbol = (NamedTypeSymbol)boundType.Type;
+                        var boundAttribute = new ExecutableCodeBinder(attributeSyntax, binder.ContainingMemberOrLambda, binder)
+                            .BindAttribute(attributeSyntax, boundTypeSymbol, this, BindingDiagnosticBag.Discarded);
+                        boundAttributeArrayBuilder.Add(boundAttribute);
+                    }
+                }
+            }
+            return boundAttributeArrayBuilder.ToImmutableAndFree();
+        }
 #nullable disable
 
         private void RecordPresenceOfBadAttributes(ImmutableArray<CSharpAttributeData> boundAttributes)

--- a/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
@@ -523,21 +523,10 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// </summary>
         protected ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder? rootBinder)
         {
-            rootBinder = rootBinder is null ? null : new ContextualAttributeBinder(rootBinder, this);
             var boundAttributeArrayBuilder = ArrayBuilder<(CSharpAttributeData, BoundAttribute)>.GetInstance();
             foreach (var attributeListSyntaxList in attributeDeclarations)
             {
-                Binder binder;
-                if (rootBinder is null)
-                {
-                    Debug.Assert(attributeListSyntaxList.Node != null);
-                    var enclosingBinder = DeclaringCompilation.GetBinderFactory(attributeListSyntaxList.Node.SyntaxTree).GetBinder(attributeListSyntaxList.Node);
-                    binder = new ContextualAttributeBinder(enclosingBinder, this);
-                }
-                else
-                {
-                    binder = rootBinder;
-                }
+                var binder = GetAttributeBinder(attributeListSyntaxList, DeclaringCompilation, rootBinder);
                 foreach (var attributeListSyntax in attributeListSyntaxList)
                 {
                     foreach (var attributeSyntax in attributeListSyntax.Attributes)
@@ -630,14 +619,9 @@ namespace Microsoft.CodeAnalysis.CSharp
 
                     if (attributesToBindCount != prevCount)
                     {
-                        Debug.Assert(attributeDeclarationSyntaxList.Node != null);
                         Debug.Assert(bindersBuilder != null);
 
-                        var syntaxTree = attributeDeclarationSyntaxList.Node.SyntaxTree;
-                        var binder = rootBinderOpt ?? compilation.GetBinderFactory(syntaxTree).GetBinder(attributeDeclarationSyntaxList.Node);
-
-                        binder = new ContextualAttributeBinder(binder, this);
-                        Debug.Assert(!binder.InAttributeArgument || this is MethodSymbol { MethodKind: MethodKind.LambdaMethod or MethodKind.LocalFunction }, "Possible cycle in attribute binding");
+                        var binder = GetAttributeBinder(attributeDeclarationSyntaxList, compilation, rootBinderOpt);
 
                         for (int i = 0; i < attributesToBindCount - prevCount; i++)
                         {
@@ -658,6 +642,16 @@ namespace Microsoft.CodeAnalysis.CSharp
                 return ImmutableArray<AttributeSyntax>.Empty;
             }
         }
+
+#nullable enable
+        private Binder GetAttributeBinder(SyntaxList<AttributeListSyntax> attributeDeclarationSyntaxList, CSharpCompilation compilation, Binder? rootBinder = null)
+        {
+            var binder = rootBinder ?? compilation.GetBinderFactory(attributeDeclarationSyntaxList.Node!.SyntaxTree).GetBinder(attributeDeclarationSyntaxList.Node);
+            binder = new ContextualAttributeBinder(binder, this);
+            Debug.Assert(!binder.InAttributeArgument || this is MethodSymbol { MethodKind: MethodKind.LambdaMethod or MethodKind.LocalFunction }, "Possible cycle in attribute binding");
+            return binder;
+        }
+#nullable disable
 
         private static bool MatchAttributeTarget(IAttributeTargetSymbol attributeTarget, AttributeLocation symbolPart, AttributeTargetSpecifierSyntax targetOpt, BindingDiagnosticBag diagnostics)
         {

--- a/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
@@ -521,10 +521,10 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <summary>
         /// Binds attributes applied to this symbol.
         /// </summary>
-        protected ImmutableArray<BoundAttribute> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder rootBinder)
+        protected ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder rootBinder)
         {
             var binder = new ContextualAttributeBinder(rootBinder, this);
-            var boundAttributeArrayBuilder = ArrayBuilder<BoundAttribute>.GetInstance();
+            var boundAttributeArrayBuilder = ArrayBuilder<(CSharpAttributeData, BoundAttribute)>.GetInstance();
             foreach (var attributeListSyntaxList in attributeDeclarations)
             {
                 foreach (var attributeListSyntax in attributeListSyntaxList)
@@ -533,8 +533,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                     {
                         var boundType = binder.BindType(attributeSyntax.Name, BindingDiagnosticBag.Discarded);
                         var boundTypeSymbol = (NamedTypeSymbol)boundType.Type;
-                        var boundAttribute = new ExecutableCodeBinder(attributeSyntax, binder.ContainingMemberOrLambda, binder)
-                            .BindAttribute(attributeSyntax, boundTypeSymbol, this, BindingDiagnosticBag.Discarded);
+                        var boundAttribute = binder.GetAttribute(attributeSyntax, boundTypeSymbol,
+                            beforeAttributePartBound: null, afterAttributePartBound: null, BindingDiagnosticBag.Discarded);
                         boundAttributeArrayBuilder.Add(boundAttribute);
                     }
                 }

--- a/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Symbol_Attributes.cs
@@ -521,12 +521,23 @@ namespace Microsoft.CodeAnalysis.CSharp
         /// <summary>
         /// Binds attributes applied to this symbol.
         /// </summary>
-        protected ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder rootBinder)
+        protected ImmutableArray<(CSharpAttributeData, BoundAttribute)> BindAttributes(OneOrMany<SyntaxList<AttributeListSyntax>> attributeDeclarations, Binder? rootBinder)
         {
-            var binder = new ContextualAttributeBinder(rootBinder, this);
+            rootBinder = rootBinder is null ? null : new ContextualAttributeBinder(rootBinder, this);
             var boundAttributeArrayBuilder = ArrayBuilder<(CSharpAttributeData, BoundAttribute)>.GetInstance();
             foreach (var attributeListSyntaxList in attributeDeclarations)
             {
+                Binder binder;
+                if (rootBinder is null)
+                {
+                    Debug.Assert(attributeListSyntaxList.Node != null);
+                    var enclosingBinder = DeclaringCompilation.GetBinderFactory(attributeListSyntaxList.Node.SyntaxTree).GetBinder(attributeListSyntaxList.Node);
+                    binder = new ContextualAttributeBinder(enclosingBinder, this);
+                }
+                else
+                {
+                    binder = rootBinder;
+                }
                 foreach (var attributeListSyntax in attributeListSyntaxList)
                 {
                     foreach (var attributeSyntax in attributeListSyntax.Attributes)

--- a/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
@@ -5691,6 +5691,30 @@ class C
         }
 
         [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
+        public void LocalMethod_AttributeArguments_GenericParameter()
+        {
+            var source = """
+                using System;
+                class A : Attribute
+                {
+                    public A(int param) { }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        [A(default(T))] void F<T>() { }
+                        F<int>();
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (10,20): error CS0246: The type or namespace name 'T' could not be found (are you missing a using directive or an assembly reference?)
+                //         [A(default(T))] void F<T>() { }
+                Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "T").WithArguments("T").WithLocation(10, 20));
+        }
+
+        [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
         public void LocalMethod_AttributeArguments_StringInterpolation()
         {
             var source = """

--- a/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
@@ -5570,12 +5570,16 @@ class C
                     static void Main()
                     {
                         const int N = 10;
+                        const int Unused = 20;
                         void F([Optional, DefaultParameterValue(N)] int x) => Console.WriteLine(x);
                         F();
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,19): warning CS0219: The variable 'Unused' is assigned but its value is never used
+                //         const int Unused = 20;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Unused").WithArguments("Unused").WithLocation(8, 19));
         }
 
         [Fact, WorkItem(63911, "https://github.com/dotnet/roslyn/issues/63911")]
@@ -5589,12 +5593,16 @@ class C
                     static void Main()
                     {
                         const int N = 10;
+                        const int Unused = 20;
                         var lam = ([Optional, DefaultParameterValue(N)] int x) => Console.WriteLine(x);
                         lam(100);
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (8,19): warning CS0219: The variable 'Unused' is assigned but its value is never used
+                //         const int Unused = 20;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Unused").WithArguments("Unused").WithLocation(8, 19));
         }
 
         [Fact, WorkItem(63911, "https://github.com/dotnet/roslyn/issues/63911")]
@@ -5612,12 +5620,16 @@ class C
                     static void Main()
                     {
                         const int N = 10;
+                        const int Unused = 20;
                         void F([A(Prop = N)] int x) { }
                         F(100);
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (12,19): warning CS0219: The variable 'Unused' is assigned but its value is never used
+                //         const int Unused = 20;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Unused").WithArguments("Unused").WithLocation(12, 19));
         }
 
         [Fact, WorkItem(63911, "https://github.com/dotnet/roslyn/issues/63911")]
@@ -5635,16 +5647,20 @@ class C
                     static void Main()
                     {
                         const int N = 10;
+                        const int Unused = 20;
                         var lam = ([A(Prop = N)] int x) => { };
                         lam(100);
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (12,19): warning CS0219: The variable 'Unused' is assigned but its value is never used
+                //         const int Unused = 20;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Unused").WithArguments("Unused").WithLocation(12, 19));
         }
 
-        [Fact, WorkItem(1614551, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1614551")]
-        public void LocalMethod_Attribute()
+        [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
+        public void LocalMethod_AttributeArguments()
         {
             var source = """
                 using System;
@@ -5661,16 +5677,51 @@ class C
                         const int N2 = 20;
                         const int N3 = 30;
                         const int N4 = 40;
-                        [A(N1, Prop = N2)][return: A(N3, Prop = N4)] int F() => 50;
+                        const int N5 = 50;
+                        const int N6 = 60;
+                        [A(N1, Prop = N2)][return: A(N3, Prop = N4)] int F() => N5;
                         F();
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (16,19): warning CS0219: The variable 'N6' is assigned but its value is never used
+                //         const int N6 = 60;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "N6").WithArguments("N6").WithLocation(16, 19));
         }
 
-        [Fact, WorkItem(1614551, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1614551")]
-        public void LambdaMethod_Attribute()
+        [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
+        public void LocalMethod_AttributeArguments_StringInterpolation()
+        {
+            var source = """
+                public class C
+                {
+                    public int P
+                    {
+                        get
+                        {
+                            const string X = "Hello";
+                            const string Y = "World";
+                            const string Z = "unused";
+                            [My($"{X}, World", Prop = $"Hello, {Y}")] int F() => 0;
+                            return F();
+                        }
+                    }
+                }
+                public class MyAttribute : System.Attribute
+                {
+                    public MyAttribute(string param) { }
+                    public string Prop { get; set; }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (9,26): warning CS0219: The variable 'Z' is assigned but its value is never used
+                //             const string Z = "unused";
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Z").WithArguments("Z").WithLocation(9, 26));
+        }
+
+        [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
+        public void LambdaMethod_AttributeArguments()
         {
             var source = """
                 using System;
@@ -5687,12 +5738,47 @@ class C
                         const int N2 = 20;
                         const int N3 = 30;
                         const int N4 = 40;
-                        var lam = [A(N1, Prop = N2)][return: A(N3, Prop = N4)] () => 50;
+                        const int N5 = 50;
+                        const int N6 = 60;
+                        var lam = [A(N1, Prop = N2)][return: A(N3, Prop = N4)] () => N5;
                         lam();
                     }
                 }
                 """;
-            CreateCompilation(source).VerifyDiagnostics();
+            CreateCompilation(source).VerifyDiagnostics(
+                // (16,19): warning CS0219: The variable 'N6' is assigned but its value is never used
+                //         const int N6 = 60;
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "N6").WithArguments("N6").WithLocation(16, 19));
+        }
+
+        [Fact, WorkItem(60645, "https://github.com/dotnet/roslyn/issues/60645")]
+        public void LambdaMethod_AttributeArguments_StringInterpolation()
+        {
+            var source = """
+                public class C
+                {
+                    public int P
+                    {
+                        get
+                        {
+                            const string X = "Hello";
+                            const string Y = "World";
+                            const string Z = "unused";
+                            var f = [My($"{X}, World", Prop = $"Hello, {Y}")] () => 0;
+                            return f();
+                        }
+                    }
+                }
+                public class MyAttribute : System.Attribute
+                {
+                    public MyAttribute(string param) { }
+                    public string Prop { get; set; }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics(
+                // (9,26): warning CS0219: The variable 'Z' is assigned but its value is never used
+                //             const string Z = "unused";
+                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Z").WithArguments("Z").WithLocation(9, 26));
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
@@ -4,6 +4,8 @@
 
 #nullable disable
 
+using System.Collections.Immutable;
+using System.Linq;
 using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
@@ -5803,6 +5805,47 @@ class C
                 // (9,26): warning CS0219: The variable 'Z' is assigned but its value is never used
                 //             const string Z = "unused";
                 Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "Z").WithArguments("Z").WithLocation(9, 26));
+        }
+
+        [Fact]
+        public void Setter_AttributeArguments()
+        {
+            var source = """
+                using System;
+                class A : Attribute
+                {
+                    public A(int param) { }
+                    public int Prop { get; set; }
+                }
+                class Program
+                {
+                    const int N1 = 10;
+                    const int N2 = 20;
+                    const int N3 = 30;
+                    const int N4 = 40;
+                    const int N5 = 50;
+                    const int N6 = 60;
+                    public int Prop
+                    {
+                        [A(N1, Prop = N4)][param: A(N2, Prop = N5)][return: A(N3, Prop = N6)]
+                        set
+                        {
+                            Console.WriteLine(value);
+                        }
+                    }
+                }
+                """;
+            var compilation = CreateCompilation(source).VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+            var declarations = tree.GetRoot().DescendantNodes().OfType<PropertyDeclarationSyntax>().ToImmutableArray();
+            var property = declarations.Select(d => model.GetDeclaredSymbol(d)).Where(p => p.ContainingSymbol.Name == "Program").Single();
+            var parameter = property.SetMethod.Parameters[0].GetSymbol<SourceComplexParameterSymbolBase>();
+            var attributes = parameter.BindParameterAttributes();
+            Assert.Equal(3, attributes.Length);
+            Assert.Equal("A(10, Prop = 40)", attributes[0].Item1.ToString());
+            Assert.Equal("A(20, Prop = 50)", attributes[1].Item1.ToString());
+            Assert.Equal("A(30, Prop = 60)", attributes[2].Item1.ToString());
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
+++ b/src/Compilers/CSharp/Test/Emit2/FlowAnalysis/FlowTests.cs
@@ -5642,5 +5642,57 @@ class C
                 """;
             CreateCompilation(source).VerifyDiagnostics();
         }
+
+        [Fact, WorkItem(1614551, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1614551")]
+        public void LocalMethod_Attribute()
+        {
+            var source = """
+                using System;
+                class A : Attribute
+                {
+                    public A(int param) { }
+                    public int Prop { get; set; }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        const int N1 = 10;
+                        const int N2 = 20;
+                        const int N3 = 30;
+                        const int N4 = 40;
+                        [A(N1, Prop = N2)][return: A(N3, Prop = N4)] int F() => 50;
+                        F();
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(1614551, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1614551")]
+        public void LambdaMethod_Attribute()
+        {
+            var source = """
+                using System;
+                class A : Attribute
+                {
+                    public A(int param) { }
+                    public int Prop { get; set; }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        const int N1 = 10;
+                        const int N2 = 20;
+                        const int N3 = 30;
+                        const int N4 = 40;
+                        var lam = [A(N1, Prop = N2)][return: A(N3, Prop = N4)] () => 50;
+                        lam();
+                    }
+                }
+                """;
+            CreateCompilation(source).VerifyDiagnostics();
+        }
     }
 }

